dual/dual.go

```diff
@@ -0,0 +1,280 @@
+	helper "github.com/libp2p/go-libp2p-routing-helpers"
+)
+
+// DHT implements the routing interface to provide two concrete DHT implementationts for use
+// in IPFS that are used to support both global network users and disjoint LAN usecases.
+type DHT struct {
+	WAN *dht.IpfsDHT
+	LAN *dht.IpfsDHT
+}
+
+// DefaultLanExtension is used to differentiate local protocol requests from those on the WAN DHT.
+const DefaultLanExtension protocol.ID = "/lan"
+
+// Assert that IPFS assumptions about interfaces aren't broken. These aren't a
+// guarantee, but we can use them to aid refactoring.
+var (
+	_ routing.ContentRouting = (*DHT)(nil)
+	_ routing.Routing        = (*DHT)(nil)
+	_ routing.PeerRouting    = (*DHT)(nil)
+	_ routing.PubKeyFetcher  = (*DHT)(nil)
+	_ routing.ValueStore     = (*DHT)(nil)
+)
+
+// New creates a new DualDHT instance. Options provided are forwarded on to the two concrete
+// IpfsDHT internal constructions, modulo additional options used by the Dual DHT to enforce
+// the LAN-vs-WAN distinction.
+// Note: query or routing table functional options provided as arguments to this function
+// will be overriden by this constructor.
+func New(ctx context.Context, h host.Host, options ...dht.Option) (*DHT, error) {
+	wanOpts := append(options,
+		dht.QueryFilter(dht.PublicQueryFilter),
+		dht.RoutingTableFilter(dht.PublicRoutingTableFilter),
+	)
+	wan, err := dht.New(ctx, h, wanOpts...)
+	if err != nil {
+		return nil, err
+	}
+
+	// Unless overridden by user supplied options, the LAN DHT should default
+	// to 'AutoServer' mode.
+	lanOpts := append(options,
+		dht.ProtocolExtension(DefaultLanExtension),
+		dht.QueryFilter(dht.PrivateQueryFilter),
+		dht.RoutingTableFilter(dht.PrivateRoutingTableFilter),
+	)
+	if wan.Mode() != dht.ModeClient {
+		lanOpts = append(lanOpts, dht.Mode(dht.ModeServer))
+	}
+	lan, err := dht.New(ctx, h, lanOpts...)
+	if err != nil {
+		return nil, err
+	}
+
+	impl := DHT{wan, lan}
+	return &impl, nil
+}
+
+// Close closes the DHT context.
+func (dht *DHT) Close() error {
+	return mergeErrors(dht.WAN.Close(), dht.LAN.Close())
+}
+
+func (dht *DHT) activeWAN() bool {
+	return dht.WAN.RoutingTable().Size() > 0
+}
+
+// Provide adds the given cid to the content routing system.
+func (dht *DHT) Provide(ctx context.Context, key cid.Cid, announce bool) error {
+	if dht.activeWAN() {
+		return dht.WAN.Provide(ctx, key, announce)
+	}
+	return dht.LAN.Provide(ctx, key, announce)
+}
+
+// FindProvidersAsync searches for peers who are able to provide a given key
+func (dht *DHT) FindProvidersAsync(ctx context.Context, key cid.Cid, count int) <-chan peer.AddrInfo {
+	reqCtx, cancel := context.WithCancel(ctx)
+	outCh := make(chan peer.AddrInfo)
+	wanCh := dht.WAN.FindProvidersAsync(reqCtx, key, count)
+	lanCh := dht.LAN.FindProvidersAsync(reqCtx, key, count)
+	go func() {
+		defer cancel()
+		defer close(outCh)
+
+		found := make(map[peer.ID]struct{}, count)
+		nch := 2
+		var pi peer.AddrInfo
+		for nch > 0 && count > 0 {
```

Comment:
another way to do this:

```go
for count > 0 && wanCh != nil && lanCh != nil {
}
```

Up to you.

Reply:
If `count == 0`, we're in "find all" mode and should keep finding till we the user cancels.